ProjFS.Mac/PrjFSKextTests/HandleOperationTests.mm

```diff
@@ -0,0 +1,569 @@
+    testFileVnode->attrValues.va_flags = FileFlags_IsEmpty | FileFlags_IsInVirtualizationRoot;
+    SetFileXattrData(testFileVnode);
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_WRITE_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_HydrateFile,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyFilePreConvertToFull,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+}
+
+-(void) testWriteFileHydrated {
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    // If we have FileXattrData attribute we should trigger MessageType_KtoU_NotifyFilePreConvertToFull to remove it
+    SetFileXattrData(testFileVnode);
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_WRITE_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertFalse(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_HydrateFile,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+    XCTAssertTrue(
+       MockCalls::DidCallFunction(
+            ProviderMessaging_TrySendRequestAndWaitForResponse,
+            _,
+            MessageType_KtoU_NotifyFilePreConvertToFull,
+            testFileVnode.get(),
+            _,
+            _,
+            _,
+            _,
+            _,
+            nullptr));
+}
+
+-(void) testWriteFileFull {
+    testFileVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot;
+    HandleVnodeOperation(
+        nullptr,
+        nullptr,
+        KAUTH_VNODE_WRITE_DATA,
+        reinterpret_cast<uintptr_t>(context),
+        reinterpret_cast<uintptr_t>(testFileVnode.get()),
+        0,
+        0);
+    XCTAssertFalse(MockCalls::DidCallFunction(ProviderMessaging_TrySendRequestAndWaitForResponse));
+}
+
+- (void) testEventsAreIgnored {
+    testDirVnode->attrValues.va_flags = FileFlags_IsInVirtualizationRoot | FileFlags_IsEmpty;
+    const int actionCount = 5;
+    kauth_action_t actions[actionCount] =
+    {
+        KAUTH_VNODE_APPEND_DATA,
+        KAUTH_VNODE_ADD_SUBDIRECTORY,
+        KAUTH_VNODE_WRITE_SECURITY,
+        KAUTH_VNODE_TAKE_OWNERSHIP,
+        KAUTH_VNODE_ACCESS
```

Comment:
Can we give `KAUTH_VNODE_ACCESS` its own test?

The kext should be ignoring all events with `KAUTH_VNODE_ACCESS` set, even those it would otherwise care about (`KAUTH_VNODE_ACCESS` is a flag that is set with other access bits (e.g. `KAUTH_VNODE_DELETE` when `HandleVnodeOperation` is called).